### PR TITLE
Add support to customize HPA name

### LINKED ISSUE #3057: Add support to customize HPA name
### Proposal

Add new field to allow custom hpa name (and not just `keda-hpa-<name>`)  

### Use-Case

For example: 
```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  annotations:
    ...
  labels:
    ...
  name: ...
spec:
  advanced:
    horizontalPodAutoscalerConfig:
      hpaName: my-prsonal-hpa-name
```

Then keda will generate hpa named `my-prsonal-hpa-name`

### Anything else?

If you're thinking it's a good proposal, I would like to try implement it myself and open PR 🙏

### PR DESCRIPTION
**Pull request by aviadlevy**
_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #3057  
Docs PR https://github.com/kedacore/keda-docs/pull/768

### CODE REVIEW

**Review by tomkerkhove — APPROVED** [review]

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:363]: I know this is an edge case, but what would happen if?
1. I do `kubectl apply` of SO with `hpaName=foo`
2. SO & HPA gets correctly created
3. KEDA Operator is scaled down for some reason
4. I do `kubectl apply` of SO with `hpaName=bar`
5. KEDA Operator is again running and will try to reconcile on the updated SO in the cluster

Comment (aviadlevy) [reply controllers/keda/scaledobject_controller.go:363]: I think that since the hpaName is saved in the SO status, when KEDA operator running again, it'll behave exactly the same:  
- Search for HPA with the name from SO status
- if not equal to the new hpa name -> delete the old hpa

Do you have any reason to believe that the SO.status is override while the KEDA operator is down?  

If you think it's necessary, I can try produce this steps you mentioned later when I'll be able to work on it

**Comment by aviadlevy** [comment]
> @aviadlevy you should also run `make build` to regenerate ScaledObject resources, as this PR adds a new field.

No git changes after `make build`  

```shell
keda on  main [⇡] via 🐹 v1.17.10 ➜  make build     
/Users/aviadl/Code/keda/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X=github.com/kedacore/keda/v2/version.GitCommit=64647753313420a59b0c0c2f764e8347c9770dce -X=github.com/kedacore/keda/v2/version.Version=main" -o bin/keda main.go
GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X=github.com/kedacore/keda/v2/version.GitCommit=64647753313420a59b0c0c2f764e8347c9770dce -X=github.com/kedacore/keda/v2/version.Version=main" -o bin/keda-adapter adapter/main.go
```

What am I missing? What did you expected to see?

**Comment by zroubalik** [comment]
> > @aviadlevy you should also run `make build` to regenerate ScaledObject resources, as this PR adds a new field.
> 
> No git changes after `make build`
> 
> ```shell
> keda on  main [⇡] via 🐹 v1.17.10 ➜  make build     
> /Users/aviadl/Code/keda/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
> go fmt ./...
> go vet ./...
> GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X=github.com/kedacore/keda/v2/version.GitCommit=64647753313420a59b0c0c2f764e8347c9770dce -X=github.com/kedacore/keda/v2/version.Version=main" -o bin/keda main.go
> GO111MODULE=on CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X=github.com/kedacore/keda/v2/version.GitCommit=64647753313420a59b0c0c2f764e8347c9770dce -X=github.com/kedacore/keda/v2/version.Version=main" -o bin/keda-adapter adapter/main.go
> ```
> 
> What am I missing? What did you expected to see?

What about `make generate` and `make manifests`?

The new field should be present in the CRD yaml file.

File: controllers/keda/scaledobject_controller.go
Comment (aviadlevy) [reply controllers/keda/scaledobject_controller.go:363]: So I had a chance to run it and I think I have bigger issue in this section (not just for the use case you described, but even on normal behaviour)  

I saw that even though the bottom line is that the hpa is created as expected, there are many errors along the way when trying to update the old hpa for no reason.

This is the output of step 5 of your use case, but now I see that this probably will happen on any change even without down scale the operator  

```
{"level":"info","ts":1653560500.109992,"logger":"controller.clustertriggerauthentication","msg":"Starting workers","reconciler group":"keda.sh","reconciler kind":"ClusterTriggerAuthentication","worker count":1}
{"level":"error","ts":1653560600.9423842,"logger":"controller.scaledobject","msg":"Failed to update HPA","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","HPA.Namespace":"default","HPA.Name":"foo","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).ensureHPAForScaledObjectExists\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:390\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).reconcileScaledObject\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:231\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).Reconcile\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:182\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560603.795263,"logger":"controller.scaledobject","msg":"Failed to check HPA for possible update","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).reconcileScaledObject\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:231\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).Reconcile\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:182\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560612.2319481,"logger":"controller.scaledobject","msg":"Failed to ensure HPA is correctly created for ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560622.908048,"logger":"controller.scaledobject","msg":"Reconciler error","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1653560624.291129,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560624.2912672,"logger":"controller.scaledobject","msg":"Creating a new HPA","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","HPA.Namespace":"default","HPA.Name":"bar"}
{"level":"info","ts":1653560624.3321779,"logger":"controller.scaledobject","msg":"Initializing Scaling logic according to ScaledObject Specification","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560624.341697,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560662.044199,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}

```

I'll try take a further look later.  
Initial thought is that I need to create the new hpa here https://github.com/kedacore/keda/pull/3068/files#diff-191f56932c20d9d4cd383f1d52fd6091f4f734988623d4b345e75b2cf9fde7d3R166   
What do you think?

File: controllers/keda/hpa.go
Comment (zroubalik) [inline controllers/keda/hpa.go:139]: Let's move this into `createAndDeployNewHPA()` once there is successful HPA creation. This function should just return HPA object.

File: controllers/keda/hpa.go
Comment (zroubalik) [inline controllers/keda/hpa.go:160]: TBH the implementation is quite confusing. We should have probably a separate check for this right in the reconcile loop. And have separate states:
1. Create a new HPA
2. Update existing HPA (no rename)
3. Rename HPA (create new and delete one)

File: controllers/keda/scaledobject_controller.go
Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:363]: Yeah, I think that we should chose a different approach in the implementation. I have had a hard times following the flow during the review. And I have concerns about the behavior, thus my original question.

Comment (zroubalik) [reply controllers/keda/scaledobject_controller.go:363]: In the meantime I did another review, see my comment: https://github.com/kedacore/keda/pull/3068#discussion_r882540907

File: controllers/keda/hpa.go
Comment (zroubalik) [reply controllers/keda/hpa.go:160]: IMHO, the main logic should be here: https://github.com/kedacore/keda/blob/26edd4f551f8f2d6a4d3ddad862c77f757ebac84/controllers/keda/scaledobject_controller.go#L362

**Comment by aviadlevy** [comment]
> 

`make generate` did the trick, though it seems like too many changes: https://github.com/kedacore/keda/pull/3068/commits/b48e8758f1253854aebdab4d6e1093628d4794d2  
@zroubalik Do you want me to revert it and leave only the new fields that added?

File: apis/keda/v1alpha1/scaledobject_types.go
Comment (aviadlevy) [reply apis/keda/v1alpha1/scaledobject_types.go]: Changed here: https://github.com/kedacore/keda/pull/3068/commits/cfb0c3884c5b6a4ce24f17baff8dc99eda0176b1

File: controllers/keda/hpa.go
Comment (aviadlevy) [reply controllers/keda/hpa.go]: Changed the flow. Now it's indeed seems better: https://github.com/kedacore/keda/pull/3068/commits/196e8f6bb76b9fe8487789e6b6b217eead474a5d

**Comment by zroubalik** [comment]
> `make generate` did the trick, though it seems like too many changes: [16b2a01](https://github.com/kedacore/keda/commit/16b2a01593c7f11f0cb29eb48b958eb310738fd2) @zroubalik Do you want me to revert it and leave only the new fields that added?

Would be nice if you could keep here in the PR only relevant changes. Would you mind opening a new PR with the extra files? We need to update them anyway. Thanks!

**Comment by aviadlevy** [comment]
> > `make generate` did the trick, though it seems like too many changes: [16b2a01](https://github.com/kedacore/keda/commit/16b2a01593c7f11f0cb29eb48b958eb310738fd2) @zroubalik Do you want me to revert it and leave only the new fields that added?
> 
> Would be nice if you could keep here in the PR only relevant changes. Would you mind opening a new PR with the extra files? We need to update them anyway. Thanks!

Kept only relevant changes.  
Once this MR will be merged I'll open MR for general CRD changes

**Comment by zroubalik** [comment]
We should have e2e test for this, covering multiple scenarios. It could be a ScaledObject with some simple scaler.

Atm the e2e tests are implemented in TS, we are doing a migration to Go. So would be nice to base that on the new implementation once the PR is merged: https://github.com/kedacore/keda/pull/3079

**Comment by aviadlevy** [comment]
FYI my last push happened during the [github action issues ](https://www.githubstatus.com/incidents/8xxms3q3qzgp) and therefore the checks didn't run on my last push. How can I re-trigger them?

**Comment by zroubalik** [comment]
@JorTurFer any idea why don't we see the rest of the checks in this PR?

**Comment by JorTurFer** [comment]
I'll close and open it again, IDK where they are... I can see it directly in actions view 
image

File: controllers/keda/hpa.go
Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: the problem is that the current behavior  in this usecase is that on step 2, KEDA already changes the labels so it won't be enough.  
IMHO this is the right behavior, i.e. KEDA operator should take ownership on existing HPA and change it / delete and re-create it if necessary  
This behavior together with this new feature will ease developers running k8s with HPA to migrate to KEDA without having to delete manually the "old" HPAs they had

Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: > the problem is that the current behavior  in this usecase is that on step 2, KEDA already changes the labels so it won't be enough.  
> IMHO this is the right behavior, i.e. KEDA operator should "take ownership" on existing Hpa with same expected name and change it / delete and re-create it if necessary  
> This behavior together with this new feature will ease developers running k8s with HPA to migrate to KEDA without having to delete manually the "old" HPAs they had

Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: @zroubalik , let me know what you think

Comment (zroubalik) [reply controllers/keda/hpa.go:188]: @aviadlevy I can see your point, but I really don't we (KEDA) should manage resources it doesn't own.

We can also check owner reference on the HPA.

Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: If this is the expected behavior, we need first to change the current behavior.  
Currently when:  
1. setting HPA named `keda-hpa-<soName>`
2. setting ScaledObject named `<soName>`
3. KEDA is taking control on the previous created HPA, adding all the labels and also adding the ownerReference.  

**Further more:**

4. when deleting the scaled object -> HPA is deleted as well

So there are some I/S changes need to be done in the current behavior to match the behavior you expecting

@zroubalik , Do you think this PR is blocked until the current behavior is changed?

Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: What I described before is a case where the user (not KEDA) created the HPA first with the name KEDA will use, and then the scaled object created, which mean that currently (without this PR) KEDA is already changing and managing the user resources (and therefore checking the owner reference won't help here as well)

Comment (zroubalik) [reply controllers/keda/hpa.go:188]: Ah gotcha. Atm KEDA doesn't care about ownership and replaces the HPA. I think that we should eventually iterate to state where KEDA checks the HPA ownership.

It might be in a separate PR though. Do you think you could try to tackle this as well? It would be really great.

Comment (aviadlevy) [reply controllers/keda/hpa.go:188]: I agree it should be on a different PR. Not sure yet if I can take it (but I'll know in the upcoming days).  

The question is whether we must first fix this or since there is no degradation we can merge this PR?

Also, can you please open an issue describing what need to be done regarding the hpa ownership? I think you'll describe what needs to be done better than me

**Comment by aviadlevy** [comment]
@zroubalik , so I presume we'll wait first for the implementation of the "HPA Ownership" ([discussed here](https://github.com/kedacore/keda/pull/3068#discussion_r890068258))
Do we have issue opened (not sure yet if I'll have the time to do it myself, but maybe someone else from the community will step up)

**Comment by zroubalik** [comment]
> @zroubalik , so I presume we'll wait first for the implementation of the "HPA Ownership" ([discussed here](https://github.com/kedacore/keda/pull/3068#discussion_r890068258)) before merging Do we have issue opened (not sure yet if I'll have the time to do it myself, but maybe someone else from the community will step up)

Yeah agree, thanks, sorry for the slow response, super busy recently ;)

**Comment by aviadlevy** [comment]
I think https://github.com/kedacore/keda/issues/3154 will cover what's needed

**Comment by v-shenoy** [comment]
#3079

> We should have e2e test for this, covering multiple scenarios. It could be a ScaledObject with some simple scaler.
> 
> Atm the e2e tests are implemented in TS, we are doing a migration to Go. So would be nice to base that on the new implementation once the PR is merged: #3079

#3079 has been merged. You can rebase onto main, and add e2e tests in `go` using the base implementation.

**Comment by aviadlevy** [comment]
> > We should have e2e test for this, covering multiple scenarios. It could be a ScaledObject with some simple scaler.
> > Atm the e2e tests are implemented in TS, we are doing a migration to Go. So would be nice to base that on the new implementation once the PR is merged: #3079
> 
> #3079 has been merged. You can rebase onto main and add e2e tests in `go` using the base implementation.

Thanks! will do!

**Comment by aviadlevy** [comment]
> > We should have e2e test for this, covering multiple scenarios. It could be a ScaledObject with some simple scaler.
> > Atm the e2e tests are implemented in TS, we are doing a migration to Go. So would be nice to base that on the new implementation once the PR is merged: #3079
> 
> #3079 has been merged. You can rebase onto main and add e2e tests in `go` using the base implementation.

Hi @v-shenoy , my e2e test does not involve any scaler in particular, only general behavior.  
Where should I locate it?

**Comment by v-shenoy** [comment]
Just make a new package inside `tests/scalers_go`


Example - `tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go`


Replace `custom_hpa_name` with something else if you prefer.

**Comment by zroubalik** [comment]
Let's use some simple scaler for this,that doesn't require big setup. @JorTurFer might suggest some

**Comment by aviadlevy** [comment]
Added e2e tests 😎 
Who can review them? @zroubalik @v-shenoy

File: tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go
Comment (JorTurFer) [inline tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go:165]: Should this be a helper? I mean, maybe other tests requires it in the future

Comment (aviadlevy) [reply tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go:165]: Wondered that myself. 
In the current e2e tests (even the tests that still written in typescript) it's not necessary, and that's why I didn't put it the for now

You think it should be in helper? I can move it to there

**Comment by aviadlevy** [comment]
> /run-e2e custom_hpa* **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/2510279672)

I increased the time waiting for hpa in e2e.

locally the e2e is working (see below), but probably on the github runner it takes more time for things to work?

```shell
➜  go test -v ./scalers_go/custom_hpa_name/custom_hpa_name_test.go
=== RUN   TestCustomToDefault
    custom_hpa_name_test.go:136: --- validate hpa is with custom name ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:140: --- change hpa name ---
    custom_hpa_name_test.go:144: --- validate new hpa is with default name ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:148: --- validate hpa with custom name does not exists ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
--- PASS: TestCustomToDefault (14.00s)
=== RUN   TestDefaultToCustom
    custom_hpa_name_test.go:136: --- validate hpa is with default name ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:140: --- change hpa name ---
    custom_hpa_name_test.go:144: --- validate new hpa is with custom name ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:148: --- validate hpa with default name does not exists ---
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
    custom_hpa_name_test.go:171: Waiting for hpa creation
--- PASS: TestDefaultToCustom (12.36s)
PASS
ok      command-line-arguments  26.894s
```

File: tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go
Comment (zroubalik) [inline tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go:12]: Could you please split imports into 3 blocks, as it is done here: https://github.com/kedacore/keda/blob/2b3fdc0e4d0c6283524be1110ef793934efee3fa/tests/scalers_go/azure_keyvault_queue/azure_keyvault_queue_test.go#L7-L23

@v-shenoy any idea why this hasn't been catched by the linter?

**Review by zroubalik — COMMENTED** [review]
Looking good! Just a nit in imports.

File: tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go
Comment (v-shenoy) [reply tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go]: I think it's because of the build tag. 

https://golangci-lint.run/usage/configuration/ 
Think we need to add

```yaml
run:
  build-tags:
  - e2e
```

**Comment by aviadlevy** [comment]
@zroubalik @JorTurFer 
Can you please define what's missing for this MR to get merged?

**Comment by JorTurFer** [comment]
> Can you please define what's missing for this MR to get merged?

I think is ready, at least from my side I don't see any other thing to be done, let's wait till @zroubalik too

**Review by JorTurFer — APPROVED** [review]

File: tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go
Comment (JorTurFer) [reply tests/scalers_go/custom_hpa_name/custom_hpa_name_test.go]: I have opened [a PR](https://github.com/kedacore/keda/pull/3189) adding this, thanks for the tip 🙇

**Review by zroubalik — APPROVED** [review]

**Comment by zroubalik** [comment]
There's a conflict brought by another PR :( could you please fix that? Then we can merge this PR. Thanks for the contribution!

**Comment by JorTurFer** [comment]
thanks for this contribution! ❤️

**Comment by aviadlevy** [comment]
> thanks for this contribution! ❤️

Thanks. it was fun and very enriching.  
Looking forward to contribute more in the future  😎

**Pull request merged**
